File: src/main/java/com/kgaft/VulkanLib/Device/Descriptors/DescriptorSet.java
package com.kgaft.VulkanLib.Device.Descriptors;

import com.kgaft.VulkanLib.Device.LogicalDevice.LogicalDevice;
import com.kgaft.VulkanLib.Utils.LwjglObject;
import org.lwjgl.vulkan.VkCommandBuffer;
import org.lwjgl.vulkan.VkWriteDescriptorSet;

import java.lang.instrument.IllegalClassFormatException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.atomic.AtomicInteger;
import static org.lwjgl.vulkan.VK13.*;
public class DescriptorSet {
    protected long[] descriptorSets;
    protected LogicalDevice device;
    private List<DescriptorBufferInfo> buffersInfo = new ArrayList<>();
    private List<DescriptorImageInfo> imagesInfo = new ArrayList<>();
    private List<DescriptorAccelerationStructureInfo> asInfo = new ArrayList<>();
    private int imageInfoPerInstanceAmount = 0;
    private int bufferInfoPerInstanceAmount = 0;

    private int imgCount = 0;
    private int buffCount = 0;
    protected DescriptorSet(){}

    public void addBufferInfo(DescriptorBufferInfo bufferInfo) {
        buffersInfo.add(bufferInfo);
    }
    public void updateDescriptors() throws IllegalClassFormatException {
        LwjglObject<VkWriteDescriptorSet.Buffer> writes = packsDescriptorWrites();
        for (long item : descriptorSets) {

            while(writes.get().hasRemaining()){
                writes.get().get().dstSet(item);
            }
            writes.get().rewind();
            if (imageInfoPerInstanceAmount == 0 && bufferInfoPerInstanceAmount == 0) {
                vkUpdateDescriptorSets(device.getDevice(), writes.get(), null);
            } else {
                vkUpdateDescriptorSets(device.getDevice(), writes.get(), null);
                writes = packsDescriptorWrites();
            }
        }
    }

    public void addImageInfo(DescriptorImageInfo imageInfo) {
        imagesInfo.add(imageInfo);
    }

    public void addAccelerationStructureInfo(DescriptorAccelerationStructureInfo acsInfo) {
        asInfo.add(acsInfo);
    }

    public void clearObjectInfos(){
        buffersInfo.clear();
        imagesInfo.clear();
        asInfo.clear();
    }

    public void bindDescriptor(int bindPoint, int currentInstance, VkCommandBuffer cmd,
                                       long pipelineLayout) {
        long[] dsc = new long[]{descriptorSets[currentInstance]};
        vkCmdBindDescriptorSets(cmd, bindPoint, pipelineLayout, 0, dsc, null);
    }

    private LwjglObject<VkWriteDescriptorSet.Buffer> packsDescriptorWrites() throws IllegalClassFormatException {
        if(imageInfoPerInstanceAmount == 0 && bufferInfoPerInstanceAmount == 0){
            LwjglObject<VkWriteDescriptorSet.Buffer> descriptorWrites = new LwjglObject<>(VkWriteDescriptorSet.class, VkWriteDescriptorSet.Buffer.class, buffersInfo.size()+imagesInfo.size()+asInfo.size());
            AtomicInteger counter = new AtomicInteger();
            buffersInfo.forEach(element->{
                descriptorWrites.get().get(counter.get()).sType$Default();
                descriptorWrites.get().get(counter.get()).dstSet(descriptorSets[0]);
                descriptorWrites.get().get(counter.get()).dstBinding(element.getBinding());
                descriptorWrites.get().get(counter.get()).dstArrayElement(0);
                descriptorWrites.get().get(counter.get()).descriptorType(element.getDescriptorType());
                descriptorWrites.get().get(counter.get()).descriptorCount(element.getBase().get().capacity());
                descriptorWrites.get().get(counter.get()).pBufferInfo(element.getBase().get());
                counter.getAndIncrement();
            });
            imagesInfo.forEach(element->{
                descriptorWrites.get().get(counter.get()).sType$Default();
                descriptorWrites.get().get(counter.get()).dstSet(descriptorSets[0]);
                descriptorWrites.get().get(counter.get()).dstBinding(element.getBinding());
                descriptorWrites.get().get(counter.get()).dstArrayElement(0);
                descriptorWrites.get().get(counter.get()).descriptorType(element.getDescriptorType());
                descriptorWrites.get().get(counter.get()).descriptorCount(element.getBase().get().capacity());
                descriptorWrites.get().get(counter.get()).pImageInfo(element.getBase().get());
                counter.getAndIncrement();
            });
            asInfo.forEach(element->{
                descriptorWrites.get().get(counter.get()).sType$Default();
                descriptorWrites.get().get(counter.get()).dstSet(descriptorSets[0]);
                descriptorWrites.get().get(counter.get()).dstBinding(element.getBinding());
                descriptorWrites.get().get(counter.get()).dstArrayElement(0);
                descriptorWrites.get().get(counter.get()).descriptorType(element.getDescriptorType());
                descriptorWrites.get().get(counter.get()).descriptorCount(1);
                descriptorWrites.get().get(counter.get()).pNext(element.getBase().get());
                counter.getAndIncrement();
            });
            return descriptorWrites;
        } else {
            if(buffCount==bufferInfoPerInstanceAmount*descriptorSets.length){
                buffCount = 0;
            }
            if(imgCount==imageInfoPerInstanceAmount*descriptorSets.length){
                imgCount = 0;
            }
            LwjglObject<VkWriteDescriptorSet.Buffer> res = new LwjglObject<>(VkWriteDescriptorSet.class, VkWriteDescriptorSet.Buffer.class, imageInfoPerInstanceAmount+bufferInfoPerInstanceAmount);
            int c = 0;
            int startBuff = buffCount;
            for(; buffCount<startBuff+bufferInfoPerInstanceAmount; buffCount++){
                res.get().get(c).sType$Default();
                res.get().get(c).dstSet(descriptorSets[0]);
                res.get().get(c).dstBinding(buffersInfo.get(buffCount).getBinding());
                res.get().get(c).dstArrayElement(0);
                res.get().get(c).descriptorType(buffersInfo.get(buffCount).getDescriptorType());
                res.get().get(c).descriptorCount(buffersInfo.get(buffCount).getBase().get().capacity());
                res.get().get(c).pBufferInfo(buffersInfo.get(buffCount).getBase().get());
                c++;
            }
            int startImg = imgCount;
            for(; imgCount<startImg+bufferInfoPerInstanceAmount; imgCount++){
                res.get().get(c).sType$Default();
                res.get().get(c).dstSet(descriptorSets[0]);
                res.get().get(c).dstBinding(imagesInfo.get(imgCount).getBinding());
                res.get().get(c).dstArrayElement(0);
                res.get().get(c).descriptorType(imagesInfo.get(imgCount).getDescriptorType());
                res.get().get(c).descriptorCount(imagesInfo.get(imgCount).getBase().get().capacity());
                res.get().get(c).pImageInfo(imagesInfo.get(imgCount).getBase().get());
                c++;
            }
            return res;
        }

    }

    public void setImageInfoPerInstanceAmount(int imageInfoPerInstanceAmount) {
        this.imageInfoPerInstanceAmount = imageInfoPerInstanceAmount;
    }

    public void setBufferInfoPerInstanceAmount(int bufferInfoPerInstanceAmount) {
        this.bufferInfoPerInstanceAmount = bufferInfoPerInstanceAmount;
    }
}

File: src/cxx/VulkanLib/Device/Descriptors/DescriptorSet.cpp
//
// Created by kgaft on 3/11/24.
//
#include "DescriptorSet.hpp"

void DescriptorSet::updateDescriptors() {

    uint32_t counter = 0;
    for (auto &item: descriptorSet) {

        for (auto &cItem: writes) {
            cItem.dstSet = item;
        }
        if (imageInfoPerInstanceAmount == 0 && bufferInfoPerInstanceAmount == 0) {
            device->getDevice().updateDescriptorSets((uint32_t)writes.size(), writes.data(), 0, nullptr);
        } else {
            device->getDevice().updateDescriptorSets(imageInfoPerInstanceAmount + bufferInfoPerInstanceAmount,
                                                     &writes[counter], 0, nullptr);
        }
        counter += imageInfoPerInstanceAmount + bufferInfoPerInstanceAmount;
    }
}

void DescriptorSet::clearObjectsInfos() {
    for (auto &item: imagesInfo) {
        delete item;
    }
    imagesInfo.clear();
    for (const auto &item: buffersInfo) {
        delete item;
    }
    buffersInfo.clear();
    asInfo.clear();
    writes.clear();
}

void DescriptorSet::bindDescriptor(vk::PipelineBindPoint bindPoint, uint32_t currentInstance, vk::CommandBuffer cmd,
                                   vk::PipelineLayout pipelineLayout) {
    cmd.bindDescriptorSets(bindPoint, pipelineLayout, 0, 1, &descriptorSet[currentInstance], 0, nullptr);
}

void DescriptorSet::addBufferInfo(DescriptorBufferInfo &bufferInfo) {
    buffersInfo.push_back(new DescriptorBufferInfo(bufferInfo));
    writes.push_back({});
    writes[writes.size() - 1].sType = vk::StructureType::eWriteDescriptorSet;
    writes[writes.size() - 1].dstSet = descriptorSet[0];
    writes[writes.size() - 1].dstBinding = buffersInfo[buffersInfo.size() - 1]->binding;
    writes[writes.size() - 1].dstArrayElement = 0;
    writes[writes.size() - 1].descriptorType = buffersInfo[buffersInfo.size() - 1]->descriptorType;
    writes[writes.size() - 1].descriptorCount =(uint32_t) buffersInfo[buffersInfo.size() - 1]->base.size();
    writes[writes.size() - 1].pBufferInfo = buffersInfo[buffersInfo.size() - 1]->base.data();
}

void DescriptorSet::addImageInfo(DescriptorImageInfo &imageInfo) {
    imagesInfo.push_back(new DescriptorImageInfo(imageInfo));
    writes.push_back({});
    writes[writes.size() - 1].sType = vk::StructureType::eWriteDescriptorSet;
    writes[writes.size() - 1].dstSet = descriptorSet[0];
    writes[writes.size() - 1].dstBinding = imagesInfo[imagesInfo.size() - 1]->binding;
    writes[writes.size() - 1].dstArrayElement = 0;
    writes[writes.size() - 1].descriptorType = imagesInfo[imagesInfo.size() - 1]->descriptorType;
    writes[writes.size() - 1].descriptorCount = (uint32_t)imagesInfo[imagesInfo.size() - 1]->imageInfos.size();
    writes[writes.size() - 1].pImageInfo = imagesInfo[imagesInfo.size() - 1]->imageInfos.data();
}

void DescriptorSet::addAccelerationStructureInfo(DescriptorAccelerationStructureInfo &acsInfo) {
    asInfo.push_back(new DescriptorAccelerationStructureInfo(acsInfo));
    writes.push_back({});

    writes[writes.size() - 1].sType = vk::StructureType::eWriteDescriptorSet;
    writes[writes.size() - 1].dstSet = descriptorSet[0];
    writes[writes.size() - 1].dstBinding = asInfo[asInfo.size() - 1]->binding;
    writes[writes.size() - 1].dstArrayElement = 0;
    writes[writes.size() - 1].descriptorType = asInfo[asInfo.size() - 1]->descriptorType;
    writes[writes.size() - 1].descriptorCount = 1;
    writes[writes.size() - 1].pNext = &asInfo[asInfo.size() - 1]->base;
}

void DescriptorSet::setImageInfoPerInstanceAmount(uint32_t imageInfoPerInstanceAmount) {
    DescriptorSet::imageInfoPerInstanceAmount = imageInfoPerInstanceAmount;
}

void DescriptorSet::setBufferInfoPerInstanceAmount(uint32_t bufferInfoPerInstanceAmount) {
    DescriptorSet::bufferInfoPerInstanceAmount = bufferInfoPerInstanceAmount;
}
